### PR TITLE
Amir full lung sliding update

### PR DESCRIPTION
**Pull request by AmirRoshankar**
DBUP and WandB integration as well as pleural line and lung point experiments

### CODE REVIEW

File: src/data/dataset_creator.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+
+        with open(query_path, 'r') as query_file:
+            clips_df = pd.read_sql(query_file.read(), conn)
+
+        # Close ssh connection
+        server.stop()
+
+        # Add a column indicating whether the sample is lung sliding or not
+        is_absent_lung_sliding = ((clips_df['pleural_line_findings'] == 'absent_lung_sliding') 
+                                    | (clips_df['pleural_line_findings'] == 'thickened|absent_lung_sliding'))
```
Comment (bvanberl) [inline src/data/dataset_creator.py:244]: should we just change this to something like if `'absent_lung_sliding' in clips_df['pleural_line_findings']`? not sure of correct syntax for that though

File: src/data/dataset_creator.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    if perform_external_val:
+        # Load a different dataset for each external location
+        for ext in external_locations.keys():
+            print(f"Setting up dataset for external location: {ext}")
+            
+            # Create lung sliding dataset creator
+            dataset_creator = LungSlidingDatasetCreator(cfg,
+                                                        database_cfg,
+                                                        external = ext)
+            # dataset_creator.create_dataset()
```
Comment (bvanberl) [inline src/data/dataset_creator.py:361]: delete comment

File: src/data/fast_pl_detector.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    :param oa: first vector from point o to a
+    :param ob: second vector from point o to b
+    :return: angle between the vectors
+    '''
+    cosine_angle = np.dot(oa, ob) / (np.linalg.norm(oa) * np.linalg.norm(ob))
+    angle = np.arccos(cosine_angle)
+    return angle
+
+
+def shadow_ave(img : NDArray,
```
Comment (bvanberl) [inline src/data/fast_pl_detector.py:79]: maybe we could rename to make this more intuitive?

File: src/data/fast_pl_detector.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+def radon_tform(img : NDArray) -> float:
+    '''
+    Uses the radon transform to calculate by what angle to rotate the image to get A-Lines
+    to be horizontal
+    :param img: LUS image (average or single frame)
+    :return: angle of rotation
+    '''
+    img = rescale(img, scale=0.3, mode='reflect', channel_axis=None)
+    smallest_angle = 70.0
```
Comment (bvanberl) [inline src/data/fast_pl_detector.py:137]: should we define these as global constants? i.e. outside of the functions 
`SMALLEST_ANGLE = 70.0`

File: src/preprocessor.py [hunk 5]
```diff
@@ -292,21 +265,26 @@ def parse_tf_m_mode(png_path, label):
...
     :param png_path: Path to LUS clip to load
     :param label: Binary label indicating whether the LUS clip exhibits lung sliding
-
     :return: Tuple of (M-mode reconstruction, label), both as tensors
     '''
     # first decode the given png_path into a string
     png_path = png_path.decode("utf-8")
     mmode_image = cv2.imread(png_path, cv2.IMREAD_GRAYSCALE)
+    mmode_image = cv2.cvtColor(mmode_image, cv2.COLOR_GRAY2RGB)
+    # assert(mmode_image.shape == (224, 180, 3))
```
Comment (bvanberl) [inline src/preprocessor.py:276]: can we delete these?

**Review by bvanberl — APPROVED** [review]

File: src/preprocessor.py [hunk 6]
```diff
@@ -367,22 +342,19 @@ def __init__(self, preprocessing_fn, batch_size=None):
...
         :param shuffle: A boolean to decide if shuffling is desired
         :param augment: A boolean to decide if augmentation is desired
-
         :return: A TF dataset with either preprocessed data or a full pipeline for eventual preprocessing
         '''
 
         # Shuffle the dataset
         if shuffle:
-            shuffle_val = len(df)
+            shuffle_val = len(ds)
```
Comment (joshuakurien) [inline src/preprocessor.py:357]: dataset.cardinality().numpy() might be better to use this instead of len. See [here](https://stackoverflow.com/questions/50737192/tf-data-dataset-how-to-get-the-dataset-size-number-of-elements-in-an-epoch/63487037#63487037)

File: src/data/dataset_creator.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+    if num_neg < num_pos:
+        negative_df = minority_upsample(negative_df, num_pos)
+    else:
+        positive_df = minority_upsample(positive_df, num_neg)
+    
+    balanced = pd.concat([negative_df, positive_df])
+    return balanced
+
+
+class LungSlidingDatasetCreator(DatasetCreator):
```
Comment (joshuakurien) [inline src/data/dataset_creator.py:78]: Solid implementation!

**Review by joshuakurien — COMMENTED** [review]
Great work Amir! Left a few minor comments.

File: src/data/auto_masking.py [hunk 8]
```diff
@@ -23,7 +23,7 @@ def get_bounding_box(self, binary_mask):
         return [min(i), max(i) , min(j), max(j)]
 
     def predict(self, input_paths, output_path, model_path,
```
[the remark this answers is not in the document]
Comment (AmirRoshankar) [reply src/data/auto_masking.py:25]: We're only using this for the smaller extra experiments which I haven't had time to integrate with the rest of the project (and with DBUP). All the main experiments use DBUP for masking though.

File: src/gradcam.py [hunk 9]
```diff
@@ -33,11 +35,11 @@ def __init__(self):
...
-    def get_preprocessed_data(self, df):
+    def get_preprocessed_data(self, df : pd.DataFrame) -> tf.data.Dataset:
         '''
         Samples df based on proportion in the config file and returns it with preprocessing
         :param df: DataFrame with data to preprocess
-        :returns: Preprocessed data
+        :return: Preprocessed data
         '''
         prop = cfg['EXPLAINABILITY']['PROPORTION']
         if prop <= 0 or prop >= 1:
```
Comment (delaneysmith22) [inline src/gradcam.py:45]: Should be prop > 1 (allow for prop = 1). Should also edit this part of the code so that if config['EXPLAINABILITY']['SINGLE_MMODE'] = True, prop is automatically set to 1. Currently, unless the user changes the config, an error is returned.

File: src/gradcam.py [hunk 10]
```diff
@@ -86,7 +88,8 @@ def __init__(self):
                 layer_name = layer.name
         self.last_conv_layer = layer_name
 
-    def apply_gradcam(self, df=pd.read_csv(cfg['EXPLAINABILITY']['PATHS']['PNG_DF'])):
+    def apply_gradcam(self,
```
Comment (delaneysmith22) [inline src/gradcam.py:91]: Line 106 currently throws an error. predict_set() current has a non-optional argument called frames_dir that is not being passed here. Either frames_dir should be made into an optional argument, or this line needs to be updated.

File: src/explanations.py [hunk 11]
```diff
@@ -9,12 +9,13 @@
 from tensorflow.keras.models import load_model
```
Comment (delaneysmith22) [inline src/explanations.py:9]: Why do we have a script explanations.py **and** a script gradcam.py? We should condense into 1 file unless they serve distinct purposes. If they do, it should be more clear from their names or outlined in the readme

File: src/lung_point_experiment.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+    mmode_image = np.clip(mmode_image,a_min=0,a_max=255)
+
+    mmode_image = mmode_image.reshape((cfg['PREPROCESS']['PARAMS']['M_MODE_HEIGHT'], 
+                                       cfg['PREPROCESS']['PARAMS']['M_MODE_WIDTH'], 
+                                       1))
+
+    return mmode_image
+
+
+def vid_to_mmode(masked_path : str,
```
Comment (delaneysmith22) [inline src/lung_point_experiment.py:55]: Same idea as above, redundant with the functions from to_png.py

File: src/lung_point_experiment_setup.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
+import shutil
```
Comment (delaneysmith22) [inline src/lung_point_experiment_setup.py:1]: How does this script differ from lung_point_experiment.py? There seems to be redundant functions and their individual purposes (if different) do not seem clear to me

File: src/predict.py [hunk 14]
```diff
@@ -4,21 +4,27 @@
...
 from custom.metrics import Specificity
 import time
 
 cfg = yaml.full_load(open(os.path.join(os.getcwd(), 'config.yml'), 'r'))
 
 
-def predict_set(model, test_df, threshold=0.5, model_def_str=cfg['TRAIN']['MODEL_DEF']):
+def predict_set(model : Model,
+                test_df : pd.DataFrame,
+                frames_dir : str,
```
Comment (delaneysmith22) [inline src/predict.py:25]: I mentioned this above, but you added this frames_dir argument here but did not update the reference to this function in gradcam.py. Please double check if it needs to be updated in other files as well.

File: src/external_validation.py [hunk 15]
```diff
@@ -199,36 +159,7 @@ def setup_experiment(experiment_path=None):
     return trial_path
 
 
-def sample_folds(clip_df, trial_folder, k, seed=None):
```
[the remark this answers is not in the document]
Comment (AmirRoshankar) [reply src/external_validation.py:202]: Hey yeah I deleted since the logging.py script in DBUP has this method in it: log_k_fold_TAAFT
It should be in the deep breathe utils repo if you'd like to take a look
